test/js-api/limits.any.js

```diff
@@ -167,30 +150,81 @@ testLimit("function returns", 0, kJSEmbeddingMaxFunctionReturns, (builder, count
         const type = builder.addType({params: [], results: array});
     });
 
-testLimit("initial table size", 1, kJSEmbeddingMaxTableSize, (builder, count) => {
-        builder.setTableBounds(count, undefined);
-    });
-
-testLimit("maximum table size", 1, kJSEmbeddingMaxTableSize, (builder, count) => {
-        builder.setTableBounds(1, count);
-    });
-
 testLimit("element segments", 1, kJSEmbeddingMaxElementSegments, (builder, count) => {
         builder.setTableBounds(1, 1);
         const array = [];
         for (let i = 0; i < count; i++) {
-            builder.addElementSegment(0, false, array, false);
+            builder.addElementSegment(0, false, false, array);
         }
     });
 
 testLimit("tables", 0, kJSEmbeddingMaxTables, (builder, count) => {
-        for (let i = 0; i < count; i++) {
-            builder.addImportedTable("", "", 1, 1);
-        }
-    });
+  for (let i = 0; i < count; i++) {
+    builder.addImportedTable("", "", 1, 1);
+  }
+});
 
 testLimit("memories", 0, kJSEmbeddingMaxMemories, (builder, count) => {
-        for (let i = 0; i < count; i++) {
-            builder.addImportedMemory("", "", 1, 1, false);
-        }
-    });
+  for (let i = 0; i < count; i++) {
+    builder.addImportedMemory("", "", 1, 1, false);
+  }
+});
+
+// This function runs the {gen} function with the value {limit+1}. Even with
+// this value, validation and compilation should succeed. An error should only
+// happen during instantiation or runtime.
+function testDynamicLimit(name, limit, test_instantiation, gen) {
+  function get_buffer(count) {
+    const builder = new WasmModuleBuilder();
+    gen(builder, count);
+    return builder.toBuffer();
+  }
+
+  const buffer = get_buffer(limit + 1);
+
+  test(() => { assert_true(WebAssembly.validate(buffer)); },
+       `Validate ${name} beyond its dynamic limit`);
+
+  test(() => { new WebAssembly.Module(buffer); },
+       `Compile ${name} beyond its dynamic limit`);
+
+  promise_test(t => { return WebAssembly.compile(buffer); },
+               `Async compile ${name} beyond its dynamic limit.`);
+
+  if (test_instantiation) {
+    test(() => {
+      const compiled_module = new WebAssembly.Module(buffer);
+      assert_throws(new WebAssembly.RuntimeError(),
+                    () => new WebAssembly.Instance(compiled_module, {}));
+    }, `Instantiate ${name} over limit`);
+
+    promise_test(t => {
+      return promise_rejects(t, new WebAssembly.RuntimeError(),
+                             WebAssembly.instantiate(buffer, {}));
+    }, `Async instantiate ${name} over limit`);
+  }
+}
+
+testDynamicLimit(
+    "initial declared memory pages", kJSEmbeddingMaxMemoryPages, true,
+    (builder, count) => { builder.addMemory(count, undefined, false, false); });
```

Comment:
We should also test the case where a memory is initially within the limit, but then is grown past the limit, including with the `memory.grow` instruction in wasm proper, and with `WebAssembly.Memory.grow()`.

However, note that the spec already requires a `RangeError` if `mem_grow` fails. It doesn't seem well-defined whether `RangeError` or `RuntimeError` wins here.

Reply:
I added tests for growing now. I changed the expectation to RangeError.